stdnum/at/fibunr.py

```diff
@@ -0,0 +1,115 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301 USA
+"""Firmenbuchnummer (Austrian Company register numbers)
+The commercial register is a public register in Austria , in which the most
+important legal facts and legal relationships are to be entered by companies
+and other institutions subject to company law. The number consists of 7
+charicters, first 6 chars are numbers and one last control letter.
+The number is based on modulo 17 where the reminder is translated into letter
+according to a know dictionary.
+Refernce: http://www.pruefziffernberechnung.de/F/Firmenbuchnummer.shtml
+>>> _calc_checksum('092754')
+'D'
+>>> is_checksum_valid('123456D')
+True
+>>> is_checksum_valid('123456T')
+False
+>>> validate('123456D')
+'123456D'
+>>> validate('1234567D')
+Traceback (most recent call last):
+    ...
+InvalidLength: ...
+>>> validate('X23456D')
+Traceback (most recent call last):
+    ...
+InvalidFormat: ...
+>>> validate('123456T')
+Traceback (most recent call last):
+    ...
+InvalidChecksum: ...
+>>> is_valid('123456D')
+True
+>>> is_valid('123456A')
+False
+"""
+from stdnum.util import clean
+from stdnum.exceptions import (
+    InvalidChecksum,
+    InvalidFormat,
+    InvalidLength,
+    ValidationError
+)
+
+
+def compact(number):
+    """Convert the number to the minimal representation. This removes
+    surrounding whitespaces and add zeros to the left of the number until it
+    become 7 characters.
+    """
+    number = clean(number, ' ').strip()
+    number = number.rjust(7, '0')
+    return number
+
+
+def _calc_checksum(number):
+    """Calculate the control digit, excpect 6 digits number."""
+    # Each position of a letter represnt the value of that letter.
+    key_table = 'ABDFGHIKMPSTVWXYZ'
+    # Each position of a number represent the weight of its counterpart in the
+    # Firmenbuchnummer.
+    weight_table = [6, 4, 14, 15, 10, 1]
+    summation = 0
+    for char_index in range(6):
+        summation += (
+            int(number[char_index]) * int(weight_table[char_index])
+        )
+
+    return key_table[summation % 17]
+
+
+def is_checksum_valid(number):
+    """
+    Calculate the checksum from the first 6 charicters and compare it to
+    the last digit"""
+    control_letter = _calc_checksum(number[:6])
+    return number.endswith(control_letter)
+
+
+def validate(number):
+    """Checks if the number provided is valid. This checks the length, chars
+    and check digit."""
+    number = compact(number)
+
+    if len(number) > 7:
+        raise InvalidLength()
+    if not number[:-1].isdigit():
+        raise InvalidFormat('The first six characters should be numbers')
+    if not is_checksum_valid(number):
```

Comment:
what should we do if we get lower case? should we covert to uppercase or it's ok to fail?